File: 2001-3000/2381-shifting-letters-ii/2381-shifting-letters-ii.java
class Solution {
    public String shiftingLetters(String s, int[][] shifts) {
        int n = s.length();
        int arr[] = new int[n+1];
        for(int shift[]:shifts){
            int x = shift[0];
            int y = shift[1];
            int delta = shift[2]==1?1:-1;
            arr[x] += delta;
            arr[y+1] -= delta;
        }
        for(int i=1;i<n;i++) arr[i] += arr[i-1];
        char[] ch = new char[n];
        for(int i=0;i<n;i++){
            int val = (s.charAt(i)-'a'+arr[i])%26;
            if(val<0) val+=26;
            ch[i] = (char)(val+'a');
        }
        return new String(ch);
    }
}

File: 2001-3000/2381-shifting-letters-ii/2381-shifting-letters-ii.cpp
class Solution {
public:
    string shiftingLetters(string s, vector<vector<int>>& shifts) {
int m=s.size();
vector<int>diff(m+1,0);
int n=shifts.size();
for(int i=0;i<n;i++){
    if(shifts[i][2]==0){
        diff[shifts[i][0]]--;
        diff[shifts[i][1]+1]++;
    }
    else{
        diff[shifts[i][0]]++;
        diff[shifts[i][1]+1]--;
    }
}
int sum=0;
for(int i=0;i<m;i++){
    sum+=diff[i];
    int pos=s[i]-'a';
    int shift=(((pos+sum)%26)+26)%26;
    s[i]='a'+shift;
    }
    return s;
    }
};